Sync/BatchingClient.swift

```diff
@@ -0,0 +1,263 @@
+        }
+
+        // Need to deduce how big the data is we're sending across using this reducer
+        let sizeReducer: (total: ByteCount, record: Record<T>) -> ByteCount = { total, record in
+            return total + (self.serializeRecord(record) ?? "").utf8.count
+        }
+
+        let sizeOfRecords = records.reduce(0, combine: sizeReducer)
+
+        // Batch up all of the records and fail early if there is an issue
+        let batchingResult = batchesFromRecords(records)
+        guard let batches = batchingResult.successValue else {
+            let failure = batchingResult.failureValue!
+            log.debug("Unable to generate batches from records submitted to batch client: \(failure)")
+            return deferMaybe(failure)
+        }
+
+        // We have too many records for a single post so we'll either need to upload in a single batch or
+        // multiple batches
+        if records.count > config.maxTotalRecords || sizeOfRecords > config.maxTotalBytes {
+            return self.bunchUpload(batches, ifUnmodifiedSince: self.ifUnmodifiedSince, onCollectionUploaded: onCollectionUploaded)
+        }
+
+        if records.count > config.maxPostRecords || sizeOfRecords > config.maxPostBytes {
+            return self.batchUpload(batches, ifUnmodifiedSince: self.ifUnmodifiedSince, onCollectionUploaded: onCollectionUploaded)
+        }
+
+        // We can just do a single post instead of batching
+        log.debug("Batch fits within a single request. Submitting records in a single post.")
+        let lines = optFilter(records.map(self.serializeRecord))
+        return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: nil)
+            >>== effect({ onCollectionUploaded($0.value) })
+            >>> succeed
+    }
+
+    public func bunchUpload(batch: [[String]], ifUnmodifiedSince: Timestamp?, onCollectionUploaded: (POSTResult -> Void)) -> Success {
+
+        // Need to break apart the provided batch into multiple batches (AKA a bunch) so we can fit them in batch calls
+        var mutBatch = batch
+
+        /* 
+         * When talking about a 'batch' and multiple 'batches' the word becomes very confusing. In this context,
+         *
+         * A 'batch' consists of N requests to the server beginning with a batch=true call and ends with a commit=true call.
+         * A 'bunch' is many of these calls.
+         * A partialBatch is one of the N requests made within a 'batch'.
+         *
+         * Here are some typealiases to make your life easier.
+         */
+
+        typealias Batch = [[String]]
+        typealias Bunch = [Batch]
+
+        var bunch = Bunch()
+
+        while !mutBatch.isEmpty {
+            var recordCount = 0
+            var size = 0
+            var smallerBatch = Batch()
+
+            while size <= config.maxTotalBytes && recordCount <= config.maxTotalRecords && !mutBatch.isEmpty {
+                let partialBatch = mutBatch.removeFirst()
```

Comment:
Yes, I see us sending requests before `commit`, *if the server supports batching* (and perhaps even if it doesn't, if the caller is willing).

The transaction concept still applies. Think about how a database works and its interface. You call `BEGIN TRANSACTION`, and can then do whatever you like for however long you like. At some point you call `ROLLBACK` or `END TRANSACTION`, and you expect the externally visible state of the database to either solidify or return to what it was. That doesn't mean that the database isn't writing stuff to disk during the transaction — indeed, it almost certainly is!

The same applies here. When you have a batch token, it means you can write whatever you want to the server, and it doesn't get *committed* until you make a request with `commit=true`. If `commit` is never called, the batch will be abandoned and will be garbage collected. The sensible thing to do, then, is to start making those HTTP requests as soon as we can — you might be uploading megabytes of records, and it doesn't make much sense to wait until we have everything before doing so.

Perhaps the best way to think about this is in terms of Clojure-esque channels and transducers. A channel is a producer-driven queue, like a conveyor belt. A transducer is a worker between two conveyor belts. In this case, our caller is pushing records onto the conveyor belt. We have a transducer that's bundling up records into collections of 50/1MB, and one that's shifting a conveyor belt into different bunches as a limit is reached (similar process). The consumer is uploading each bundle once it arrives. Finally, you tell the destination to commit when the channel is closed.

If the server doesn't support batching, then yes, it's unsafe to leave large gaps between uploads. But we do already upload in chunks without batching support, and it's likely that more records will be delivered for upload before our first upload is complete, so the only concern would be if we aborted the entire upload operation due to a malformed record or were interrupted (partial write). That's why we built atomic uploads in the first place, so I wouldn't spend too much time worrying about that.

From the caller's perspective, the contract remains: the records it added to the batcher will be permanently committed to the server at some point between calling `commit` and the `onCollectionUploaded` callback being invoked. We just happen to sometimes move them up to the server before `commit` is called.